src/passes/freeFunctions.ts

```diff
@@ -0,0 +1,134 @@
+import assert from 'assert';
+import {
+  ContractDefinition,
+  ContractKind,
+  FunctionCall,
+  FunctionDefinition,
+  Identifier,
+  MemberAccess,
+} from 'solc-typed-ast';
+import { AST } from '../ast/ast';
+import { ASTMapper } from '../ast/mapper';
+import { cloneASTNode } from '../utils/cloning';
+
+/* 
+  Pass:
+  The pass inlines free functions into the contract if 
+  the free functions make library calls.
+*/
+
+export class FreeFunctions extends ASTMapper {
+  visitContractDefinition(node: ContractDefinition, ast: AST): void {
+    // Store all free functions
+    const freeFunctions = new Map<number, FunctionDefinition>();
```

Comment:
This is a very stateful approach and it makes it quite difficult to follow and to check for edge cases. I think most of the state can be removed with the following changes:
- Check `FunctionDefinition.kind` rather than checking whether a node is in the `freeFunctions` map, this removes the need for the `freeFunctions` map
- Instead of creating an `identifierId` map, create a list of `FunctionDefinitions` that need inlining. The inlining process can be done in a map function that returns pairs of `[newFunctionDefinition, oldFunctionDefinition]`, and for each of these the contract's children that are `Identifiers` can be iterated over and their `referencedDeclaration` and `name` can be updated. I think this replaces both `identifierIds` and `remappingIds` with a single `functionsToInline` array
- Instead of checking whether an id is in `contractDefLib`, get the node itself and check whether it's a Library. This removes the need to iterate over the entire AST and store and pass around the result